### DISCUSSION TITLE
Compilation time?

### DISCUSSION THREAD

**Discussion by RobB720** [lnis-uofu/OpenFPGA#394] [Q&A]
Hi,
My first download and compilation! In the video it's stated that compilation takes 5 minutes on a 64 CPU machine - does anyone know or can estimate how long it might take on a standard (and quite old) desktop?
Many thanks,
Rob.

**Comment by RobB720** [comment]
I have a compilations problem - using Centos7 with gcc verion 8 - I'm not sure if this is the right place to ask for help, but...
[ 89%] Building CXX object vpr/CMakeFiles/libvpr.dir/src/util/hash.cpp.o
[ 89%] Building CXX object vpr/CMakeFiles/libvpr.dir/src/util/histogram.cpp.o
[ 89%] Building CXX object vpr/CMakeFiles/libvpr.dir/src/util/vpr_error.cpp.o
[ 89%] Building CXX object vpr/CMakeFiles/libvpr.dir/src/util/vpr_utils.cpp.o
In file included from /usr/include/tbb/enumerable_thread_specific.h:32,
                 from /usr/include/tbb/combinable.h:32,
                 from /home/robb/OpenFPGA/libs/EXTERNAL/libtatum/libtatum/tatum/graph_walkers/ParallelLevelizedWalker.hpp:8,
                 from /home/robb/OpenFPGA/libs/EXTERNAL/libtatum/libtatum/tatum/graph_walkers.hpp:9,
                 from /home/robb/OpenFPGA/libs/EXTERNAL/libtatum/libtatum/tatum/analyzer_factory.hpp:10,
                 from /home/robb/OpenFPGA/vpr/src/timing/timing_info.h:6,
                 from /home/robb/OpenFPGA/vpr/src/timing/concrete_timing_info.cpp:3:
/usr/include/tbb/concurrent_vector.h: In member function ‘tbb::concurrent_vector<T, A>::iterator tbb::concurrent_vector<T, A>::grow_by(tbb::concurrent_vector<T, A>::size_type)’:
/usr/include/tbb/concurrent_vector.h:667:38: error: operands to ?: have different types ‘tbb::internal::concurrent_vector_base_v3::size_type’ {aka ‘long unsigned int’} and ‘tbb::atomic<long unsigned int>’
         return iterator(*this, delta ? internal_grow_by( delta, sizeof(T), &initialize_array, NULL ) : my_early_size);
                                ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/tbb/concurrent_vector.h:667:38: note:   and each type can be converted to the other
/usr/include/tbb/concurrent_vector.h: In member function ‘tbb::concurrent_vector<T, A>::iterator tbb::concurrent_vector<T, A>::grow_by(tbb::concurrent_vector<T, A>::size_type, tbb::concurrent_vector<T, A>::const_reference)’:
/usr/include/tbb/concurrent_vector.h:680:38: error: operands to ?: have different types ‘tbb::internal::concurrent_vector_base_v3::size_type’ {aka ‘long unsigned int’} and ‘tbb::atomic<long unsigned int>’

As its VPR, could it be related to the gtk-3 comment on the compile help page:

"VPR’s GUI requires gtk-3, and can be enabled with cmake .. -DVPR_USE_EZGL=on"

I have gtk-3 installed - below is an extract from "yum list installed":
gtk3.x86_64                            3.22.30-6.el7                   @updates 
gtk3-devel.x86_64                      3.22.30-6.el7                   @updates 
gtk3-immodule-xim.x86_64               3.22.30-6.el7                   @updates 

What is meant by "enabled"? in the compile help - where/how do I run "cmake .. -DVPR_USE_EZGL=on"?

Any help much appreciated! Very keen to get working with this tool!

Many thanks,
Rob.

**Comment by tangxifan** [comment]
@RobB720 For runtime, you can refer to our github action about build compatibility test: https://github.com/lnis-uofu/OpenFPGA/runs/3929728766

Regarding the errors, I believe it is due to some TBB issues known in VTR. My suggestion is to disable TBB as we did in CI tests. 
Regarding the GUI, if you have installed the dependencies at https://github.com/lnis-uofu/OpenFPGA/blob/master/.github/workflows/install_dependencies_build.sh, CMake will enable it automatically.

**Comment by RobB720** [comment]
Thanks for the reply!

Apologies, this is all new to me, but how do I "disable TBB as we did in CI tests"? 

For the GUI, the dependencies list includes "tcl8.6-dev" - whereas my Centos 7 installation has 8.5:

tcl.x86_64                                 1:8.5.13-8.el7         @anaconda     
tcl-devel.x86_64                           1:8.5.13-8.el7         @anaconda     

Could this be the problem? I think I can install tcl8.6 directly if needed (its not available through yum for centos 7). Going through the dependencies in more detail a few on the list do not have the same names in Centos (dev->devel) and some I can't find at all e.g. libevent-dev, libc6-dev, libcairo2-dev, libfontconfig1-dev, liblist-moreutils-perl, libncurses5-dev, libreadline-dev, libxml++2.6-dev and qt5-default

So is Centos 7 the issue here? 

I have a new installation so am happy to change my version of linux to Centos 8 or Ubuntu if that is easier.

**Comment by tangxifan** [comment]
@RobB720 If you are new, I suggest you to use the same environment as we did in Github Actions:

https://github.com/lnis-uofu/OpenFPGA/blob/5e912b3c5113bf6be3ed29563e922b6791f70896/.github/workflows/build.yml#L23

**Reply by RobB720** [reply]
OK I have installed Ubuntu 18.04.6 LTS and am running with default gcc (it seems to have v7 and v8 installed but I don't know which one is being picked up) but I get compiler errors this time:

[ 72%] Linking CXX executable rtl_number
gcc: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
[ 81%] Building techlibs/ecp5/ecp5_ffinit.o
Makefile:677: recipe for target 'passes/memory/memory_share.o' failed
make[4]: *** [passes/memory/memory_share.o] Error 4
gcc: internal compiler error: Killed (program cc1plus)
gcc: internal compiler error: Killed (program cc1plus)
make[4]: *** Waiting for unfinished jobs....

One installation requirement that I don't have is cmake > 3.12. However, Ubuntu says I have the latest 3.10.2:

sudo apt upgrade cmake
[sudo] password: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
cmake is already the newest version (3.10.2-1ubuntu2.18.04.2).
Calculating upgrade... Done
0 to upgrade, 0 to newly install, 0 to remove and 0 not to upgrade.

Any help much appreciated!
Best regards,
Rob.

**Reply by tangxifan** [reply]
If you check your compilation log, you should be able to see the gcc/g++ version you are using https://github.com/lnis-uofu/OpenFPGA/runs/3957366021?check_suite_focus=true#step:8:52

**Reply by tangxifan** [reply]
It looks like an internal error of the compiler. Suggest to change to gcc-8 or something higher. gcc-7 is sometimes buggy
